src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,4971 @@
+
+        public ReadOnlyMemory<byte> GetEnumeratedBytes(Asn1Tag expectedTag)
+        {
+            // T-REC-X.690-201508 sec 8.4 says the contents are the same as for integers.
+            ReadOnlyMemory<byte> contents =
+                GetIntegerContents(expectedTag, UniversalTagNumber.Enumerated, out int headerLength);
+
+            _data = _data.Slice(headerLength + contents.Length);
+            return contents;
+        }
+
+        public TEnum GetEnumeratedValue<TEnum>() where TEnum : struct
+        {
+            Type tEnum = typeof(TEnum);
+
+            return (TEnum)Enum.ToObject(tEnum, GetEnumeratedValue(tEnum));
+        }
+
+        public TEnum GetEnumeratedValue<TEnum>(Asn1Tag expectedTag) where TEnum : struct
+        {
+            Type tEnum = typeof(TEnum);
+
+            return (TEnum)Enum.ToObject(tEnum, GetEnumeratedValue(expectedTag, tEnum));
+        }
+
+        public Enum GetEnumeratedValue(Type tEnum) =>
+            GetEnumeratedValue(new Asn1Tag(UniversalTagNumber.Enumerated), tEnum);
+
+        public Enum GetEnumeratedValue(Asn1Tag expectedTag, Type tEnum)
+        {
+            const UniversalTagNumber tagNumber = UniversalTagNumber.Enumerated;
+            
+            // This will throw an ArgumentException if TEnum isn't an enum type,
+            // so we don't need to validate it.
+            Type backingType = tEnum.GetEnumUnderlyingType();
+
+            if (tEnum.IsDefined(typeof(FlagsAttribute), false))
+            {
+                throw new ArgumentException(
+                    SR.Cryptography_Asn_EnumeratedValueRequiresNonFlagsEnum,
+                    nameof(tEnum));
+            }
+
+            // T-REC-X.690-201508 sec 8.4 says the contents are the same as for integers.
+            int sizeLimit = Marshal.SizeOf(backingType);
+
+            if (backingType == typeof(int) ||
+                backingType == typeof(long) ||
+                backingType == typeof(short) ||
+                backingType == typeof(sbyte))
+            {
+                if (!TryReadSignedInteger(sizeLimit, expectedTag, tagNumber, out long value))
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                return (Enum)Enum.ToObject(tEnum, value);
+            }
+
+            if (backingType == typeof(uint) ||
+                backingType == typeof(ulong) ||
+                backingType == typeof(ushort) ||
+                backingType == typeof(byte))
+            {
+                if (!TryReadUnsignedInteger(sizeLimit, expectedTag, tagNumber, out ulong value))
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                return (Enum)Enum.ToObject(tEnum, value);
+            }
+
+            Debug.Fail($"No handler for type {backingType.Name}");
+            throw new CryptographicException();
```

Comment:
nit: message

Reply:
This should never happen, so no message is required.